kolena/_experimental/trace/trace.py

```diff
@@ -0,0 +1,239 @@
+        self.task_ongoing = None
+        self.sync_interval = sync_interval
+        self.record_timestamp = record_timestamp
+        atexit.register(self._clean_up)
+
+    def _add_id_fields(self, datapoint: Dict, result: Dict) -> None:
+        for field in self.id_fields:
+            if field == KOLENA_DEFAULT_ID:
+                call_id = uuid.uuid4().hex
+                datapoint[KOLENA_DEFAULT_ID] = call_id
+                result[KOLENA_DEFAULT_ID] = call_id
+            else:
+                field_value = datapoint.get(field)
+                if field_value is None:
+                    raise ValueError(f"Id Field {field} cannot be None in datapoint input")
+                result[field] = field_value
+
+    def __call__(self, *args, **kwargs):
+        bounded_arguments = self.signature.bind(*args, **kwargs)
+        bounded_arguments.apply_defaults()
+        arguments = bounded_arguments.arguments
+
+        start_time = datetime.now()
+        output = self.func(**arguments)
+        end_time = datetime.now()
+        datapoint = _serialize_item(arguments)
+        result = _serialize_item(output)
+        if not isinstance(result, dict):
+            result = {"result": result}
+        if self.record_timestamp:
+            datapoint[KOLENA_TIMESTAMP_KEY] = start_time.isoformat()
+            result[KOLENA_TIMESTAMP_KEY] = end_time.isoformat()
+            result[KOLENA_TIME_ELAPSED_KEY] = (end_time - start_time).total_seconds()
+        self._add_id_fields(datapoint, result)
+        self.datapoints.append(datapoint)
+        model_name = self.model_name
+        if self.model_name_field and arguments.get(self.model_name_field) is not None:
+            model_name = arguments.get(self.model_name_field)
+        self.results[model_name].append(result)
+        if time.time() - self.last_update > self.sync_interval and (
+            self.task_ongoing is None or not self.task_ongoing.is_alive()
+        ):
+            self.task_ongoing = threading.Thread(target=self._push_data)
+            self.task_ongoing.start()
+        return output
+
+    def _push_data(self):
+        try:
+            dataset_df = pd.DataFrame(self.datapoints)
+
+            _upload_dataset(self.dataset_name, dataset_df, id_fields=self.id_fields, append_only=True)
+            for model_name, results in self.results.items():
+                result_df = pd.DataFrame(results)
+                upload_results(self.dataset_name, model_name, result_df)
+                self.results = self.results[result_df.shape[0] :]
+            self.last_update = time.time()
+            self.datapoints = self.datapoints[dataset_df.shape[0] :]
+
+        except Exception as e:
+            print(f"Failed to sync data: {e}")
+
+    def _clean_up(self):
+        if self.task_ongoing and self.task_ongoing.is_alive():
+            self.task_ongoing.join()
+        if self.datapoints:
+            self._push_data()
+
+
+def kolena_trace(
```

Comment:
as the good practice for decorator, we might need to use [functools.wraps](https://docs.python.org/3/library/functools.html#functools.wraps) at here.

```sh

In [1]: from kolena._experimental.trace import kolena_trace

In [2]: def f():
   ...:     """this is my doc string"""
   ...:     return 42
   ...: 

In [3]: @kolena_trace
   ...: def f():
   ...:     """this is my doc string"""
   ...:     return 42
   ...: 

In [4]: f.__name__
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-c97c7504e772> in <cell line: 1>()
----> 1 f.__name__

AttributeError: '_Trace' object has no attribute '__name__'

In [5]: f.__doc__

```

see discussion like [this](https://stackoverflow.com/a/309000) as well/

Reply:
good call

Reply:
in my understanding, decorator should work as a regular function as well

```sh
In [9]: g = kolena_trace(f())
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-3a9816fd2509> in <cell line: 1>()
----> 1 g = kolena_trace(f())

~/Documents/kolena-repos/kolena/kolena/_experimental/trace/trace.py in __call__(self, *args, **kwargs)
    143         self.datapoints.append(datapoint)
    144         model_name = self.model_name
--> 145         if self.model_name_field and arguments.get(self.model_name_field) is not None:
    146             model_name = arguments.get(self.model_name_field)
    147         self.results[model_name].append(result)

AttributeError: '_Trace' object has no attribute 'model_name_field'

In [10]: g = kolena_trace(f)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-10-9d4cbe888115> in <cell line: 1>()
----> 1 g = kolena_trace(f)

~/Documents/kolena-repos/kolena/kolena/_experimental/trace/trace.py in kolena_trace(func, dataset_name, model_name, model_name_field, sync_interval, id_fields, record_timestamp)
    215     """
    216     if func:
--> 217         return _Trace(
    218             func,
    219             dataset_name=dataset_name,

~/Documents/kolena-repos/kolena/kolena/_experimental/trace/trace.py in __init__(self, func, dataset_name, model_name, model_name_field, sync_interval, id_fields, record_timestamp)
     81         self.signature = inspect.signature(func)
     82         if not dataset_name:
---> 83             dataset_name = func.__name__
     84         self.dataset_name = dataset_name
     85         if model_name:

AttributeError: '_Trace' object has no attribute '__name__'
```

Reply:
the new commit should be good, tested these

```
from kolena._experimental.trace import kolena_trace
def f2():
    return 4
f3 = kolena_trace(f2)
f3.__name__
'f2'
@kolena_trace
def f1():
    return 5
f1.__name__
'f1'
@kolena_trace(dataset_name="test")
def f4():
    return 5
f4.__name__
'f4'
```